docs/proposals/templates.md

```diff
@@ -0,0 +1,404 @@
+* Manually instantiating each template and utilizing [service linking](https://github.com/kubernetes/kubernetes/pull/17543) to share 
+  any necessary configuration data.
+
+
+## Proposed Design
+
+We begin by looking at the List object which allows a user to easily group a set of objects together for easy creation via a 
+single CLI invocation.  It also provides a portable format which requires only a single file to represent an application.
+
+From that starting point, we propose a Template API object which can encapsulate the definition of all components of an 
+application to be created.  The application definition is encapsulated in the form of an array of API objects (identical to 
+List), plus a parameterization section.  Components reference the parameter by name and the value of the parameter is 
+substituted during a processing step, prior to submitting each component to the appropriate API endpoint for creation.
+
+The primary capability provided is that parameter values can easily be shared between components, such as a database password 
+that is provided by the user once, but then attached as an environment variable to both a database pod and a web frontend pod.
+
+In addition, the template can be repeatedly instantiated for a consistent application deployment experience in different 
+namespaces or Kubernetes clusters.
+
+Lastly, we propose the Template API object include a “Labels” section in which the template author can define a set of labels 
+to be applied to all objects created from the template.  This will give the template deployer an easy way to manage all the 
+components created from a given template.
+
+### API Objects
+
+**Template Object**
+```
+// Template contains the inputs needed to produce a Config.
+type Template struct {
+    unversioned.TypeMeta
+    kapi.ObjectMeta
+
+    // Optional: Parameters is an array of Parameters used during the
+    // Template to Config transformation.
+    Parameters []Parameter
+
+    // Required: A list of resources to create
+    Objects []runtime.Object
+
+    // Optional: ObjectLabels is a set of labels that are applied to every
+    // object during the Template to Config transformation
+    ObjectLabels map[string]string
+}
+```
+
+**Parameter Object**
+```
+// Parameter defines a name/value variable that is to be processed during
+// the Template to Config transformation.
+type Parameter struct {
+    // Required: Parameter name must be set and it can be referenced in Template
+    // Items using $(PARAMETER_NAME)
+    Name string
+
+    // Optional: The name that will show in UI instead of parameter 'Name'
+    DisplayName string
+
+    // Optional: Parameter can have description
+    Description string
+
+    // Optional: Value holds the Parameter data. If specified, the generator
+    // will be ignored. The value replaces all occurrences of the Parameter
+    // $(Name) expression during the Template to Config transformation.
+    Value string
```

Comment:
Agree, again this restriction is discussed at the end of the doc and i'd love to eliminate it.  Adding a "type" field to the Parameter struct is one of the possible ways to address it, and it's the one i'm currently leaning towards.